# <center>Outlier detection and removal using percentile</center>

In [1]:
import pandas as pd

df = pd.read_csv("height.csv")

In [15]:
df.head()

name  height
0  mohan     5.9
1  maria     5.2
2  sakib     5.1
3    tao     5.5
4  virat     4.9

### Let's check maximum thresold

In [16]:
max_thresold = df['height'].quantile(0.95)
max_thresold

9.689999999999998

In [17]:
df[df['height']>max_thresold]

name  height
9  imran    14.5

this data point is showing the height of imran is 14.5 and that's not possible

### Check the minimum thresold percentile

In [18]:
min_thresold = df['height'].quantile(0.05)
min_thresold

3.6050000000000004

In [19]:
df[df['height']<min_thresold]

name  height
12  yoseph     1.2

## Removing this Outliers

In [20]:
df[(df['height']<max_thresold) & (df['height']>min_thresold)]

name  height
0     mohan     5.9
1     maria     5.2
2     sakib     5.1
3       tao     5.5
4     virat     4.9
5    khusbu     5.4
6    dmitry     6.2
7    selena     6.5
8      john     7.1
10     jose     6.1
11  deepika     5.6
13    binod     5.5

## Now use comple dataset

In [21]:
df1= pd.read_csv('Bengaluru_House_Data.csv')

In [22]:
df1.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [24]:
df1.shape

(13320, 9)

In [25]:
df1.describe()

bath       balcony         price
count  13247.000000  12711.000000  13320.000000
mean       2.692610      1.584376    112.565627
std        1.341458      0.817263    148.971674
min        1.000000      0.000000      8.000000
25%        2.000000      1.000000     50.000000
50%        2.000000      2.000000     72.000000
75%        3.000000      2.000000    120.000000
max       40.000000      3.000000   3600.000000

In [36]:
def convert_sqft_to_num(x):
    tokens = x.split('-')
    if len(tokens) == 2:
        return (float(tokens[0])+float(tokens[1]))/2
    try:
        return float(x)
    except:
        return None

In [37]:
df2 = df1.copy()
df2.total_sqft = df2.total_sqft.apply(convert_sqft_to_num)
df1 = df1[df1.total_sqft.notnull()]
df1.head(2)

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00

In [38]:
df2.loc[30]

area_type       Super built-up  Area
availability                  19-Dec
location                   Yelahanka
size                           4 BHK
society                      LedorSa
total_sqft                      2475
bath                               4
balcony                            0
price                            186
Name: 30, dtype: object

#### Add new feature called price per square feet

In [39]:
df3 = df2.copy()
df3['price_per_sqft'] = df2['price']*100000/df2['total_sqft']
df3.head()


area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society  total_sqft  bath  balcony   price  price_per_sqft  
0  Coomee       1056.0   2.0      1.0   39.07     3699.810606  
1  Theanmp      2600.0   5.0      3.0  120.00     4615.384615  
2      NaN      1440.0   2.0      3.0   62.00     4305.555556  
3  Soiewre      1521.0   3.0      1.0   95.00     6245.890861  
4      NaN      1200.0   2.0      1.0   51.00     4250.000000

## Now check the thresold

In [40]:
minimum_thresold, maximum_thresold = df3.price_per_sqft.quantile([0.001, 0.999])

In [41]:
df3[df3.price_per_sqft < minimum_thresold]

area_type          availability  \
674          Built-up  Area         Ready To Move   
810              Plot  Area                18-Apr   
1894             Plot  Area         Ready To Move   
5393   Super built-up  Area         Ready To Move   
6423             Plot  Area  Immediate Possession   
7242             Plot  Area         Ready To Move   
7947             Plot  Area         Ready To Move   
8306             Plot  Area  Immediate Possession   
8391   Super built-up  Area         Ready To Move   
9240             Plot  Area         Ready To Move   
11019            Plot  Area  Immediate Possession   
11569            Plot  Area  Immediate Possession   
11748        Built-up  Area         Ready To Move   
12470  Super built-up  Area         Ready To Move   

                              location       size  society  total_sqft  bath  \
674                          Yelahanka      3 BHK  Edreelm     35000.0   3.0   
810    4 Bedroom Farm House in Bagalur  4 Bedroom      NaN     10961.0   4.0   
1894                       Nelamangala  3 Bedroom      NaN     52272.0   2.0   
5393                  Doddabommasandra      9 BHK      NaN     42000.0   8.0   
6423                        Whitefield        NaN  SRniaGa      2324.0   NaN   
7242                         Yelahanka  1 Bedroom      NaN     26136.0   1.0   
7947                          JP Nagar      3 BHK      NaN     20000.0   3.0   
8306                            Anekal        NaN  Vaensnd      1453.0   NaN   
8391                           Kengeri      1 BHK      NaN      1200.0   1.0   
9240             SARJAPUR BAGALUR ROAD  4 Bedroom      NaN     10961.0   4.0   
11019                        Yelahanka        NaN  Saaveha      1500.0   NaN   
11569                       Hosur Road        NaN  AVeldun      1350.0   NaN   
11748                            Begur      3 BHK  SrSaira      2400.0   3.0   
12470                 Nagashetty Halli      4 BHK      NaN     16335.0   4.0   

       balcony   price  price_per_sqft  
674        3.0  130.00      371.428571  
810        1.0   80.00      729.860414  
1894       1.0  140.00      267.829813  
5393       3.0  175.00      416.666667  
6423       NaN   26.73     1150.172117  
7242       0.0  150.00      573.921028  
7947       NaN  175.00      875.000000  
8306       NaN   16.50     1135.581555  
8391       1.0   14.00     1166.666667  
9240       1.0   80.00      729.860414  
11019      NaN   12.75      850.000000  
11569      NaN    8.44      625.185185  
11748      2.0   12.00      500.000000  
12470      2.0  149.00      912.151821

you can see the price per sqfeet is between 8 to 10 rs and that is not possible

In [42]:
df3[df3.price_per_sqft>maximum_thresold]

area_type   availability             location       size  \
349              Plot  Area         18-Dec   Suragajakkanahalli  3 Bedroom   
1020           Carpet  Area  Ready To Move       Weavers Colony      1 BHK   
1122         Built-up  Area  Ready To Move  Grihalakshmi Layout  5 Bedroom   
4086             Plot  Area  Ready To Move       Sarjapur  Road  4 Bedroom   
4972         Built-up  Area  Ready To Move          Srirampuram      7 BHK   
5970             Plot  Area  Ready To Move          Mysore Road  1 Bedroom   
6421             Plot  Area         18-Sep         Bommenahalli  4 Bedroom   
7088         Built-up  Area  Ready To Move          Srirampuram      1 BHK   
7657         Built-up  Area  Ready To Move       Raghuvanahalli      1 BHK   
7883   Super built-up  Area  Ready To Move        Vasanth nagar      4 BHK   
8398   Super built-up  Area  Ready To Move    Bannerghatta Road      5 BHK   
9535             Plot  Area  Ready To Move         Indira Nagar  4 Bedroom   
11558            Plot  Area         19-Oct           Whitefield  4 Bedroom   
12443            Plot  Area  Ready To Move       Dollars Colony  4 Bedroom   

       society  total_sqft  bath  balcony   price  price_per_sqft  
349    PrhyaK         11.0   3.0      2.0    74.0    6.727273e+05  
1020       NaN        15.0   1.0      0.0    30.0    2.000000e+05  
1122       NaN        24.0   2.0      2.0   150.0    6.250000e+05  
4086       NaN         1.0   4.0      NaN   120.0    1.200000e+07  
4972       NaN         5.0   7.0      3.0   115.0    2.300000e+06  
5970       NaN        45.0   1.0      0.0    23.0    5.111111e+04  
6421   Prood G      2940.0   3.0      2.0  2250.0    7.653061e+04  
7088       NaN       650.0   1.0      3.0   500.0    7.692308e+04  
7657       NaN       425.0   1.0      1.0   750.0    1.764706e+05  
7883   LeeyaCa      2000.0   3.0      2.0  1063.0    5.315000e+04  
8398       NaN      2500.0   4.0      NaN  1400.0    5.600000e+04  
9535       NaN      2400.0   5.0      NaN  1250.0    5.208333e+04  
11558  NVaree         60.0   4.0      2.0   218.0    3.633333e+05  
12443      NaN      4350.0   8.0      NaN  2600.0    5.977011e+04

Now let's remove this outliers

In [43]:
df4 = df3[(df3.price_per_sqft<maximum_thresold) & (df3.price_per_sqft>minimum_thresold)]

Now check the data shape 

In [44]:
df4.shape

(13246, 10)

In [45]:
df4.sample(10)

area_type   availability            location       size  \
4838   Super built-up  Area  Ready To Move   Bannerghatta Road      2 BHK   
4774         Built-up  Area  Ready To Move            Kadugodi      2 BHK   
545              Plot  Area  Ready To Move  2nd Phase JP Nagar  3 Bedroom   
8060   Super built-up  Area  Ready To Move         Bilal Nagar      2 BHK   
1489         Built-up  Area  Ready To Move          Srirampura      3 BHK   
7101   Super built-up  Area  Ready To Move          Kudlu Gate      3 BHK   
3358   Super built-up  Area         18-Jul           Hosa Road      3 BHK   
3143   Super built-up  Area  Ready To Move          Whitefield      2 BHK   
8166         Built-up  Area  Ready To Move              Anekal      2 BHK   
13200            Plot  Area  Ready To Move      Defence Colony  6 Bedroom   

       society  total_sqft  bath  balcony   price  price_per_sqft  
4838   PrarePa      1130.0   2.0      2.0    78.0     6902.654867  
4774   K tlete      1100.0   2.0      1.0    45.0     4090.909091  
545        NaN      2610.0   3.0      0.0   499.0    19118.773946  
8060       NaN       650.0   1.0      1.0    20.0     3076.923077  
1489   NRnoral      1756.0   3.0      3.0   100.0     5694.760820  
7101   Brlms P      1656.0   3.0      2.0    80.0     4830.917874  
3358   Sosisic      1665.0   2.0      3.0   125.0     7507.507508  
3143   BMttaor      1280.0   2.0      2.0    52.0     4062.500000  
8166       NaN       620.0   1.0      1.0    23.5     3790.322581  
13200      NaN      8000.0   6.0      3.0  2800.0    35000.000000